File: src/conditionParser.ts
/**
 * Parses and evaluates the small condition DSL used in knowledge_base.json's
 * `example_rules[].condition`, e.g.:
 *   hand_shape == 'earth'
 *   mounts.jupiter.developed == 'strong' AND fate_line.end == 'to_jupiter'
 *   life_line.marks contains 'square'
 *   (a == 'x' OR b == 'y') AND c != 'z'
 */

type TokenType = 'STRING' | 'NUMBER' | 'BOOLEAN' | 'PATH' | 'OP' | 'AND' | 'OR' | 'CONTAINS' | 'LPAREN' | 'RPAREN';

interface Token {
  type: TokenType;
  value: string;
}

const TOKEN_RE = /'([^']*)'|(\()|(\))|(==|!=)|(\d+(?:\.\d+)?)|([A-Za-z_][A-Za-z0-9_.]*)/g;

function tokenize(input: string): Token[] {
  const tokens: Token[] = [];
  let match: RegExpExecArray | null;
  TOKEN_RE.lastIndex = 0;
  while ((match = TOKEN_RE.exec(input)) !== null) {
    const [, str, lparen, rparen, op, num, word] = match;
    if (str !== undefined) {
      tokens.push({ type: 'STRING', value: str });
    } else if (lparen) {
      tokens.push({ type: 'LPAREN', value: '(' });
    } else if (rparen) {
      tokens.push({ type: 'RPAREN', value: ')' });
    } else if (op) {
      tokens.push({ type: 'OP', value: op });
    } else if (num) {
      tokens.push({ type: 'NUMBER', value: num });
    } else if (word) {
      const upper = word.toUpperCase();
      if (upper === 'AND') tokens.push({ type: 'AND', value: word });
      else if (upper === 'OR') tokens.push({ type: 'OR', value: word });
      else if (upper === 'CONTAINS') tokens.push({ type: 'CONTAINS', value: word });
      else if (word === 'true' || word === 'false') tokens.push({ type: 'BOOLEAN', value: word });
      else tokens.push({ type: 'PATH', value: word });
    }
  }
  return tokens;
}

export type ASTNode =
  | { type: 'AND'; left: ASTNode; right: ASTNode }
  | { type: 'OR'; left: ASTNode; right: ASTNode }
  | { type: 'COMPARISON'; path: string; op: '==' | '!=' | 'contains'; value: string | number | boolean };

class Parser {
  private pos = 0;

  constructor(private readonly tokens: Token[]) {}

  parse(): ASTNode {
    const node = this.parseOr();
    if (this.pos < this.tokens.length) {
      throw new Error(`Unexpected token at position ${this.pos}: '${this.tokens[this.pos].value}'`);
    }
    return node;
  }

  private parseOr(): ASTNode {
    let node = this.parseAnd();
    while (this.peek()?.type === 'OR') {
      this.pos++;
      node = { type: 'OR', left: node, right: this.parseAnd() };
    }
    return node;
  }

  private parseAnd(): ASTNode {
    let node = this.parseTerm();
    while (this.peek()?.type === 'AND') {
      this.pos++;
      node = { type: 'AND', left: node, right: this.parseTerm() };
    }
    return node;
  }

  private parseTerm(): ASTNode {
    if (this.peek()?.type === 'LPAREN') {
      this.pos++;
      const node = this.parseOr();
      this.expect('RPAREN');
      return node;
    }
    return this.parseComparison();
  }

  private parseComparison(): ASTNode {
    const pathTok = this.expect('PATH');
    const opTok = this.peek();
    if (!opTok) throw new Error(`Expected operator after path '${pathTok.value}'`);

    let op: '==' | '!=' | 'contains';
    if (opTok.type === 'OP') {
      op = opTok.value as '==' | '!=';
      this.pos++;
    } else if (opTok.type === 'CONTAINS') {
      op = 'contains';
      this.pos++;
    } else {
      throw new Error(`Expected '==', '!=' or 'contains' after path '${pathTok.value}', got ${opTok.type}`);
    }

    const valTok = this.peek();
    if (!valTok) throw new Error(`Expected value after operator for path '${pathTok.value}'`);
    this.pos++;

    let value: string | number | boolean;
    if (valTok.type === 'STRING') value = valTok.value;
    else if (valTok.type === 'NUMBER') value = parseFloat(valTok.value);
    else if (valTok.type === 'BOOLEAN') value = valTok.value === 'true';
    else throw new Error(`Invalid value token for path '${pathTok.value}': '${valTok.value}'`);

    return { type: 'COMPARISON', path: pathTok.value, op, value };
  }

  private peek(): Token | undefined {
    return this.tokens[this.pos];
  }

  private expect(type: TokenType): Token {
    const tok = this.tokens[this.pos];
    if (!tok || tok.type !== type) {
      throw new Error(`Expected ${type} at position ${this.pos}, got ${tok ? tok.type : 'EOF'}`);
    }
    this.pos++;
    return tok;
  }
}

function getPath(obj: unknown, path: string): unknown {
  return path.split('.').reduce<unknown>((acc, key) => {
    if (acc == null || typeof acc !== 'object') return undefined;
    return (acc as Record<string, unknown>)[key];
  }, obj);
}

export function parseCondition(input: string): ASTNode {
  const tokens = tokenize(input);
  if (tokens.length === 0) throw new Error('Empty condition string');
  return new Parser(tokens).parse();
}

export function evaluateCondition(ast: ASTNode, features: Record<string, unknown>): boolean {
  switch (ast.type) {
    case 'AND':
      return evaluateCondition(ast.left, features) && evaluateCondition(ast.right, features);
    case 'OR':
      return evaluateCondition(ast.left, features) || evaluateCondition(ast.right, features);
    case 'COMPARISON': {
      const actual = getPath(features, ast.path);
      if (ast.op === 'contains') {
        if (Array.isArray(actual)) return actual.includes(ast.value);
        if (typeof actual === 'string') return actual.includes(String(ast.value));
        return false;
      }
      const isEqual = actual === ast.value;
      return ast.op === '==' ? isEqual : !isEqual;
    }
  }
}

File: chiro_engine/condition_parser.py
"""
Parst und wertet die kleine Condition-DSL aus knowledge_base.json's
`example_rules[].condition` aus, z.B.:

    hand_shape == 'earth'
    mounts.jupiter.developed == 'strong' AND fate_line.end == 'to_jupiter'
    life_line.marks contains 'square'
    (a == 'x' OR b == 'y') AND c != 'z'
"""

from __future__ import annotations

import re
from dataclasses import dataclass
from typing import Any, Union

_TOKEN_RE = re.compile(
    r"'([^']*)'"          # 1: quoted string
    r"|(\()"              # 2: lparen
    r"|(\))"              # 3: rparen
    r"|(==|!=)"           # 4: operator
    r"|(\d+(?:\.\d+)?)"   # 5: number
    r"|([A-Za-z_][A-Za-z0-9_.]*)"  # 6: word (path / keyword / boolean)
)


@dataclass(frozen=True)
class Token:
    type: str
    value: str


def _tokenize(text: str) -> list[Token]:
    tokens: list[Token] = []
    for m in _TOKEN_RE.finditer(text):
        string, lparen, rparen, op, num, word = m.groups()
        if string is not None:
            tokens.append(Token("STRING", string))
        elif lparen:
            tokens.append(Token("LPAREN", "("))
        elif rparen:
            tokens.append(Token("RPAREN", ")"))
        elif op:
            tokens.append(Token("OP", op))
        elif num:
            tokens.append(Token("NUMBER", num))
        elif word:
            upper = word.upper()
            if upper == "AND":
                tokens.append(Token("AND", word))
            elif upper == "OR":
                tokens.append(Token("OR", word))
            elif upper == "CONTAINS":
                tokens.append(Token("CONTAINS", word))
            elif word in ("true", "false"):
                tokens.append(Token("BOOLEAN", word))
            else:
                tokens.append(Token("PATH", word))
    return tokens


@dataclass(frozen=True)
class Comparison:
    path: str
    op: str  # '==' | '!=' | 'contains'
    value: Union[str, float, bool]


@dataclass(frozen=True)
class BinaryOp:
    kind: str  # 'AND' | 'OR'
    left: "ASTNode"
    right: "ASTNode"


ASTNode = Union[Comparison, BinaryOp]


class _Parser:
    def __init__(self, tokens: list[Token]):
        self._tokens = tokens
        self._pos = 0

    def parse(self) -> ASTNode:
        node = self._parse_or()
        if self._pos < len(self._tokens):
            tok = self._tokens[self._pos]
            raise ValueError(f"Unexpected token at position {self._pos}: '{tok.value}'")
        return node

    def _parse_or(self) -> ASTNode:
        node = self._parse_and()
        while self._peek() is not None and self._peek().type == "OR":
            self._pos += 1
            node = BinaryOp("OR", node, self._parse_and())
        return node

    def _parse_and(self) -> ASTNode:
        node = self._parse_term()
        while self._peek() is not None and self._peek().type == "AND":
            self._pos += 1
            node = BinaryOp("AND", node, self._parse_term())
        return node

    def _parse_term(self) -> ASTNode:
        tok = self._peek()
        if tok is not None and tok.type == "LPAREN":
            self._pos += 1
            node = self._parse_or()
            self._expect("RPAREN")
            return node
        return self._parse_comparison()

    def _parse_comparison(self) -> ASTNode:
        path_tok = self._expect("PATH")
        op_tok = self._peek()
        if op_tok is None:
            raise ValueError(f"Expected operator after path '{path_tok.value}'")

        if op_tok.type == "OP":
            op = op_tok.value
            self._pos += 1
        elif op_tok.type == "CONTAINS":
            op = "contains"
            self._pos += 1
        else:
            raise ValueError(
                f"Expected '==', '!=' or 'contains' after path '{path_tok.value}', got {op_tok.type}"
            )

        val_tok = self._peek()
        if val_tok is None:
            raise ValueError(f"Expected value after operator for path '{path_tok.value}'")
        self._pos += 1

        value: Union[str, float, bool]
        if val_tok.type == "STRING":
            value = val_tok.value
        elif val_tok.type == "NUMBER":
            value = float(val_tok.value)
        elif val_tok.type == "BOOLEAN":
            value = val_tok.value == "true"
        else:
            raise ValueError(f"Invalid value token for path '{path_tok.value}': '{val_tok.value}'")

        return Comparison(path_tok.value, op, value)

    def _peek(self) -> Token | None:
        return self._tokens[self._pos] if self._pos < len(self._tokens) else None

    def _expect(self, type_: str) -> Token:
        tok = self._peek()
        if tok is None or tok.type != type_:
            got = tok.type if tok else "EOF"
            raise ValueError(f"Expected {type_} at position {self._pos}, got {got}")
        self._pos += 1
        return tok


def _get_path(obj: Any, path: str) -> Any:
    current = obj
    for key in path.split("."):
        if not isinstance(current, dict):
            return None
        current = current.get(key)
    return current


def parse_condition(text: str) -> ASTNode:
    tokens = _tokenize(text)
    if not tokens:
        raise ValueError("Empty condition string")
    return _Parser(tokens).parse()


def evaluate_condition(ast: ASTNode, features: dict) -> bool:
    if isinstance(ast, BinaryOp):
        if ast.kind == "AND":
            return evaluate_condition(ast.left, features) and evaluate_condition(ast.right, features)
        return evaluate_condition(ast.left, features) or evaluate_condition(ast.right, features)

    actual = _get_path(features, ast.path)
    if ast.op == "contains":
        if isinstance(actual, (list, tuple, set)):
            return ast.value in actual
        if isinstance(actual, str):
            return str(ast.value) in actual
        return False

    is_equal = actual == ast.value
    return is_equal if ast.op == "==" else not is_equal
